### PR TITLE
refactors code and tests site populatecollection

### PR DESCRIPTION
**Pull request by HiteshRepo**
refactors code and tests site populatecollection

This is the final PR of the support of 'previouspath' for sharepoint lists PR chain:
- https://github.com/alcionai/corso/pull/4938
- https://github.com/alcionai/corso/pull/4946
- https://github.com/alcionai/corso/pull/4947
- https://github.com/alcionai/corso/pull/4948
- https://github.com/alcionai/corso/pull/4950

Tested manually if the deleted list in the site gets reflected in the next backup or not and here are the results:
**Backup#1**
Site has a list named `temp-list` with list id `d5b9b13e-4646-4374-902e-7c3c2cd4501d`.

lists stored in kopia (11 entries):
```
drwxrwxrwx        18959 2023-12-27 22:27:09 IST k12cbdb5cc1272f21a4bd89b80865b059  MjJlNjA5YWMtZTU0OC00MzY1LTk4ZTctM2IyYWY5Mjg5MDg2/
drwxrwxrwx        20937 2023-12-23 19:17:59 IST kb6b9f232cc3e12a4e6d409472adfa16a  NTVkMTkwY2QtNTEwOS00NzljLThlNWQtZjg4YzczNjYzMjJh/
drwxrwxrwx        17152 2023-12-14 13:00:35 IST k497bcade330714d9e319719cb69719f0  NzU0YWY3YTQtNjgyZi00MGU2LThiMGEtOWNhYTRhYzYyOGJl/
drwxrwxrwx        18285 2023-12-10 11:40:54 IST kdb724b7af6812fb00790aa9de922de85  ZGE3NmY5OGYtYjYyMS00YTU3LThjNDktYTkxNzE4ZTM4YmRm/
drwxrwxrwx        15714 2023-12-14 13:00:29 IST kedba09b1ec8dfda71fe2783a29550b7f  ZThkZjY5YWQtZDM1Ni00ZWI4LTkyMTEtZWY3Nzc2YTIxMzVk/
drwxrwxrwx        22938 2023-12-14 13:00:29 IST k3e9b6a6a02a1269839141464efb7c82b  ZjgyZjkwMGItZmM5YS00Mzc2LTlkYmMtODkwOTA5YjNhMzAy/
drwxrwxrwx        17311 2023-12-27 22:20:56 IST k89708dca3aed5d2d46bbc4aa74866cd4  MDhhMTFhZTMtZDg0ZS00ZWYxLTg4M2ItOTA5OTcyOGExYzEx/
drwxrwxrwx        17289 2023-12-28 16:36:10 IST k9fb16c9fde760b9f1e4e59a6f5b4ad2a  MTMxNDdhODctZmExNS00OTdlLTgzNTEtYmY5YjJiYTQwMWRi/
drwxrwxrwx        15678 2023-12-14 13:00:29 IST kcf4669bb29e8bf34e1a93c3c7126568d  Mzc0OTkxYzItOTE1NC00ZjdlLWI1ZjUtOTUzMjQ2MTAwMjFh/
drwxrwxrwx        15655 2023-12-14 13:00:30 IST ka887f4845d704ffb7d104c88535cf278  NDIwNzNmNGQtZTYwNS00ZmFkLTg4ZTAtNjZiMTcyNmExODM5/
drwxrwxrwx        14570 2023-12-29 13:05:03 IST kc8d1a0a5d2c0e53305e88c4f692f3e97  ZDViOWIxM2UtNDY0Ni00Mzc0LTkwMmUtN2MzYzJjZDQ1MDFk/ --> temp-list (d5b9b13e-4646-4374-902e-7c3c2cd4501d)
```

previous paths stored in kopia (11 entries):
```
{"08a11ae3-d84e-4ef1-883b-9099728a1c11":"fb8afbaa-e94c-4ea5-8a8a-24aff04d7874/sharepoint/10rqc2.sharepoint.com,72b1468e-7a6c-4622-bf83-0e2e8f009afe,2c6a13db-cf82-41d4-ab21-0d5d4ecb4481/lists/08a11ae3-d84e-4ef1-883b-9099728a1c11","13147a87-fa15-497e-8351-bf9b2ba401db":"fb8afbaa-e94c-4ea5-8a8a-24aff04d7874/sharepoint/10rqc2.sharepoint.com,72b1468e-7a6c-4622-bf83-0e2e8f009afe,2c6a13db-cf82-41d4-ab21-0d5d4ecb4481/lists/13147a87-fa15-497e-8351-bf9b2ba401db","22e609ac-e548-4365-98e7-3b2af9289086":"fb8afbaa-e94c-4ea5-8a8a-24aff04d7874/sharepoint/10rqc2.sharepoint.com,72b1468e-7a6c-4622-bf83-0e2e8f009afe,2c6a13db-cf82-41d4-ab21-0d5d4ecb4481/lists/22e609ac-e548-4365-98e7-3b2af9289086","374991c2-9154-4f7e-b5f5-95324610021a":"fb8afbaa-e94c-4ea5-8a8a-24aff04d7874/sharepoint/10rqc2.sharepoint.com,72b1468e-7a6c-4622-bf83-0e2e8f009afe,2c6a13db-cf82-41d4-ab21-0d5d4ecb4481/lists/374991c2-9154-4f7e-b5f5-95324610021a","42073f4d-e605-4fad-88e0-66b1726a1839":"fb8afbaa-e94c-4ea5-8a8a-24aff04d7874/sharepoint/10rqc2.sharepoint.com,72b1468e-7a6c-4622-bf83-0e2e8f009afe,2c6a13db-cf82-41d4-ab21-0d5d4ecb4481/lists/42073f4d-e605-4fad-88e0-66b1726a1839","55d190cd-5109-479c-8e5d-f88c7366322a":"fb8afbaa-e94c-4ea5-8a8a-24aff04d7874/sharepoint/10rqc2.sharepoint.com,72b1468e-7a6c-4622-bf83-0e2e8f009afe,2c6a13db-cf82-41d4-ab21-0d5d4ecb4481/lists/55d190cd-5109-479c-8e5d-f88c7366322a","754af7a4-682f-40e6-8b0a-9caa4ac628be":"fb8afbaa-e94c-4ea5-8a8a-24aff04d7874/sharepoint/10rqc2.sharepoint.com,72b1468e-7a6c-4622-bf83-0e2e8f009afe,2c6a13db-cf82-41d4-ab21-0d5d4ecb4481/lists/754af7a4-682f-40e6-8b0a-9caa4ac628be","d5b9b13e-4646-4374-902e-7c3c2cd4501d":"fb8afbaa-e94c-4ea5-8a8a-24aff04d7874/sharepoint/10rqc2.sharepoint.com,72b1468e-7a6c-4622-bf83-0e2e8f009afe,2c6a13db-cf82-41d4-ab21-0d5d4ecb4481/lists/d5b9b13e-4646-4374-902e-7c3c2cd4501d","da76f98f-b621-4a57-8c49-a91718e38bdf":"fb8afbaa-e94c-4ea5-8a8a-24aff04d7874/sharepoint/10rqc2.sharepoint.com,72b1468e-7a6c-4622-bf83-0e2e8f009afe,2c6a13db-cf82-41d4-ab21-0d5d4ecb4481/lists/da76f98f-b621-4a57-8c49-a91718e38bdf","e8df69ad-d356-4eb8-9211-ef7776a2135d":"fb8afbaa-e94c-4ea5-8a8a-24aff04d7874/sharepoint/10rqc2.sharepoint.com,72b1468e-7a6c-4622-bf83-0e2e8f009afe,2c6a13db-cf82-41d4-ab21-0d5d4ecb4481/lists/e8df69ad-d356-4eb8-9211-ef7776a2135d","f82f900b-fc9a-4376-9dbc-890909b3a302":"fb8afbaa-e94c-4ea5-8a8a-24aff04d7874/sharepoint/10rqc2.sharepoint.com,72b1468e-7a6c-4622-bf83-0e2e8f009afe,2c6a13db-cf82-41d4-ab21-0d5d4ecb4481/lists/f82f900b-fc9a-4376-9dbc-890909b3a302"}
```

**Backup#2**
Deleted list `temp-list` from the site

lists stored in kopia (10 entries):
```
drwxrwxrwx        15678 2023-12-14 13:00:29 IST kcf4669bb29e8bf34e1a93c3c7126568d  Mzc0OTkxYzItOTE1NC00ZjdlLWI1ZjUtOTUzMjQ2MTAwMjFh/
drwxrwxrwx        20937 2023-12-23 19:17:59 IST kb6b9f232cc3e12a4e6d409472adfa16a  NTVkMTkwY2QtNTEwOS00NzljLThlNWQtZjg4YzczNjYzMjJh/
drwxrwxrwx        17152 2023-12-14 13:00:35 IST k497bcade330714d9e319719cb69719f0  NzU0YWY3YTQtNjgyZi00MGU2LThiMGEtOWNhYTRhYzYyOGJl/
drwxrwxrwx        18285 2023-12-10 11:40:54 IST kdb724b7af6812fb00790aa9de922de85  ZGE3NmY5OGYtYjYyMS00YTU3LThjNDktYTkxNzE4ZTM4YmRm/
drwxrwxrwx        15714 2023-12-14 13:00:29 IST kedba09b1ec8dfda71fe2783a29550b7f  ZThkZjY5YWQtZDM1Ni00ZWI4LTkyMTEtZWY3Nzc2YTIxMzVk/
drwxrwxrwx        17311 2023-12-27 22:20:56 IST k89708dca3aed5d2d46bbc4aa74866cd4  MDhhMTFhZTMtZDg0ZS00ZWYxLTg4M2ItOTA5OTcyOGExYzEx/
drwxrwxrwx        17289 2023-12-28 16:36:10 IST k9fb16c9fde760b9f1e4e59a6f5b4ad2a  MTMxNDdhODctZmExNS00OTdlLTgzNTEtYmY5YjJiYTQwMWRi/
drwxrwxrwx        18959 2023-12-27 22:27:09 IST k12cbdb5cc1272f21a4bd89b80865b059  MjJlNjA5YWMtZTU0OC00MzY1LTk4ZTctM2IyYWY5Mjg5MDg2/
drwxrwxrwx        22938 2023-12-14 13:00:29 IST k3e9b6a6a02a1269839141464efb7c82b  ZjgyZjkwMGItZmM5YS00Mzc2LTlkYmMtODkwOTA5YjNhMzAy/
drwxrwxrwx        15655 2023-12-14 13:00:30 IST ka887f4845d704ffb7d104c88535cf278  NDIwNzNmNGQtZTYwNS00ZmFkLTg4ZTAtNjZiMTcyNmExODM5/
```

previous paths stored in kopia  (10 entries):
```
{"08a11ae3-d84e-4ef1-883b-9099728a1c11":"fb8afbaa-e94c-4ea5-8a8a-24aff04d7874/sharepoint/10rqc2.sharepoint.com,72b1468e-7a6c-4622-bf83-0e2e8f009afe,2c6a13db-cf82-41d4-ab21-0d5d4ecb4481/lists/08a11ae3-d84e-4ef1-883b-9099728a1c11","13147a87-fa15-497e-8351-bf9b2ba401db":"fb8afbaa-e94c-4ea5-8a8a-24aff04d7874/sharepoint/10rqc2.sharepoint.com,72b1468e-7a6c-4622-bf83-0e2e8f009afe,2c6a13db-cf82-41d4-ab21-0d5d4ecb4481/lists/13147a87-fa15-497e-8351-bf9b2ba401db","22e609ac-e548-4365-98e7-3b2af9289086":"fb8afbaa-e94c-4ea5-8a8a-24aff04d7874/sharepoint/10rqc2.sharepoint.com,72b1468e-7a6c-4622-bf83-0e2e8f009afe,2c6a13db-cf82-41d4-ab21-0d5d4ecb4481/lists/22e609ac-e548-4365-98e7-3b2af9289086","374991c2-9154-4f7e-b5f5-95324610021a":"fb8afbaa-e94c-4ea5-8a8a-24aff04d7874/sharepoint/10rqc2.sharepoint.com,72b1468e-7a6c-4622-bf83-0e2e8f009afe,2c6a13db-cf82-41d4-ab21-0d5d4ecb4481/lists/374991c2-9154-4f7e-b5f5-95324610021a","42073f4d-e605-4fad-88e0-66b1726a1839":"fb8afbaa-e94c-4ea5-8a8a-24aff04d7874/sharepoint/10rqc2.sharepoint.com,72b1468e-7a6c-4622-bf83-0e2e8f009afe,2c6a13db-cf82-41d4-ab21-0d5d4ecb4481/lists/42073f4d-e605-4fad-88e0-66b1726a1839","55d190cd-5109-479c-8e5d-f88c7366322a":"fb8afbaa-e94c-4ea5-8a8a-24aff04d7874/sharepoint/10rqc2.sharepoint.com,72b1468e-7a6c-4622-bf83-0e2e8f009afe,2c6a13db-cf82-41d4-ab21-0d5d4ecb4481/lists/55d190cd-5109-479c-8e5d-f88c7366322a","754af7a4-682f-40e6-8b0a-9caa4ac628be":"fb8afbaa-e94c-4ea5-8a8a-24aff04d7874/sharepoint/10rqc2.sharepoint.com,72b1468e-7a6c-4622-bf83-0e2e8f009afe,2c6a13db-cf82-41d4-ab21-0d5d4ecb4481/lists/754af7a4-682f-40e6-8b0a-9caa4ac628be","da76f98f-b621-4a57-8c49-a91718e38bdf":"fb8afbaa-e94c-4ea5-8a8a-24aff04d7874/sharepoint/10rqc2.sharepoint.com,72b1468e-7a6c-4622-bf83-0e2e8f009afe,2c6a13db-cf82-41d4-ab21-0d5d4ecb4481/lists/da76f98f-b621-4a57-8c49-a91718e38bdf","e8df69ad-d356-4eb8-9211-ef7776a2135d":"fb8afbaa-e94c-4ea5-8a8a-24aff04d7874/sharepoint/10rqc2.sharepoint.com,72b1468e-7a6c-4622-bf83-0e2e8f009afe,2c6a13db-cf82-41d4-ab21-0d5d4ecb4481/lists/e8df69ad-d356-4eb8-9211-ef7776a2135d","f82f900b-fc9a-4376-9dbc-890909b3a302":"fb8afbaa-e94c-4ea5-8a8a-24aff04d7874/sharepoint/10rqc2.sharepoint.com,72b1468e-7a6c-4622-bf83-0e2e8f009afe,2c6a13db-cf82-41d4-ab21-0d5d4ecb4481/lists/f82f900b-fc9a-4376-9dbc-890909b3a302"}
```

#### Does this PR need a docs update or release note?
- [x] :no_entry: No

#### Type of change

- [x] :robot: Supportability/Tests

#### Issue(s)
#4754 

#### Test Plan

- [x] :muscle: Manual
- [x] :zap: Unit test
- [x] :green_heart: E2E

### CODE REVIEW

File: src/internal/m365/collection/site/backup_test.go [hunk 1]
```diff
@@ -22,6 +29,269 @@ import (
...
+		expectNotMovedColls int
+		expectMetadataColls int
+		expectTombstoneCols int
+	}{
+		{
+			name:  "one list",
+			lists: siteMock.StubLists("one"),
+			deltaPaths: metadata.DeltaPaths{
+				"one": {
+					Delta: "one",
```
Comment (pandeyabs) [inline src/internal/m365/collection/site/backup_test.go:193]: `Delta` would not really be used right? Given lists don't support delta. Can we remove it to avoid confusion?

File: src/internal/m365/collection/site/backup_test.go [hunk 2]
```diff
@@ -22,6 +29,269 @@ import (
...
+				},
+			},
+			expectErr:           require.NoError,
+			expectColls:         3,
+			expectNewColls:      1,
+			expectMetadataColls: 1,
+			expectTombstoneCols: 1,
+		},
+		{
+			name:  "two lists, one deleted",
```
Comment (pandeyabs) [inline src/internal/m365/collection/site/backup_test.go:220]: Could you change the test name? It conflicts with the one above.

File: src/internal/m365/collection/site/backup_test.go [hunk 3]
```diff
@@ -22,6 +29,269 @@ import (
...
+				}
+			}
+
+			assert.Equal(t, test.expectNewColls, newStates, "new collections")
+			assert.Equal(t, test.expectMetadataColls, metadatas, "metadata collections")
+		})
+	}
+}
+
+func (suite *SharePointBackupUnitSuite) TestCollectLists_incremental() {
```
Comment (pandeyabs) [inline src/internal/m365/collection/site/backup_test.go:144]:
```suggestion
func (suite *SharePointBackupUnitSuite) TestPopulateListsCollections_incremental() {
```

File: src/internal/m365/collection/site/backup_test.go [hunk 4]
```diff
@@ -22,6 +29,269 @@ import (
...
+		deltaPaths          metadata.DeltaPaths
+		expectErr           require.ErrorAssertionFunc
+		expectColls         int
+		expectNewColls      int
+		expectNotMovedColls int
+		expectMetadataColls int
+		expectTombstoneCols int
+	}{
+		{
+			name:  "one list",
```
Comment (pandeyabs) [inline src/internal/m365/collection/site/backup_test.go:189]: Could you also add tests with a) empty `deltaPaths` b) unchanged collection.

File: src/internal/m365/collection/site/backup_test.go (hunk 3 above)
Comment (pandeyabs) [reply src/internal/m365/collection/site/backup_test.go:144]: Also, this takes care of missing tombstone tests. Thanks!

**Review by pandeyabs — APPROVED** [review]
Looks good!

File: src/internal/m365/collection/site/backup.go [hunk 5]
```diff
@@ -157,28 +158,57 @@ func CollectLists(
...
 	var (
-		el        = errs.Local()
-		spcs      = make([]data.BackupCollection, 0)
-		spcsMap   = make(map[string]data.BackupCollection)
-		acc       = api.CallConfig{Select: idAnd("list")}
-		currPaths = map[string]string{}
-		prevPath  path.Path
+		el   = errs.Local()
+		spcs = make([]data.BackupCollection, 0)
+		acc  = api.CallConfig{Select: idAnd("list")}
```
[the remark this answers is not in the document]
Comment (HiteshRepo) [reply src/internal/m365/collection/site/backup.go:163]: This should go away after rebase/merge. This was changed in lazy reader PRs

File: src/internal/m365/collection/site/handlers.go [hunk 6]
```diff
@@ -20,7 +20,7 @@ type backupHandler interface {
 // canonicalPath constructs the service and category specific path for
 // the given builder.
 type canonicalPather interface {
-	canonicalPath(
+	CanonicalPath(
```
[the remark this answers is not in the document]
Comment (HiteshRepo) [reply src/internal/m365/collection/site/handlers.go:23]: yeah. I have this [mock](https://github.com/alcionai/corso/blob/2a37d9b00834462ca52c1effb5f1434b853a2b0e/src/internal/m365/collection/site/mock/list.go#L67) `listHandler` that also implements `canonicalPath` and is present in a different `mock` package. We should not be changing code for tests ideally. So will add this to tech-debt

[the remark this answers is not in the document]
Comment (HiteshRepo) [reply src/internal/m365/collection/site/handlers.go:23]: I will move the mock to the same package and then I could make `canonicalPath` unexported in a different PR
